algorithms/nautilus/grammartec/recursion_info.cpp

```diff
@@ -0,0 +1,159 @@
+ * @brief Recursion of tree
+ * @author Ricerca Security <fuzzuf-dev@ricsec.co.jp>
+ *
+ * @details This file defines the RecursionInfo class.
+ *          RecursionInfo holds the discrete distribution
+ *          sampler to pick a node randomly.
+ *          It also calculates the parents of each node
+ *          and check if the parents share the same non-terminal
+ *          (recursion) as that of children.
+ */
+#include <algorithm>
+#include <iostream>
+#include <limits>
+#include <numeric>
+#include <utility>
+#include "fuzzuf/algorithms/nautilus/grammartec/recursion_info.hpp"
+#include "fuzzuf/exceptions.hpp"
+#include "fuzzuf/utils/random.hpp"
+
+
+namespace fuzzuf::algorithm::nautilus::grammartec {
+
+using fuzzuf::utils::random::WalkerDiscreteDistribution;
+
+/**
+ * @fn
+ * @brief Construct recursion info
+ * @param (t) Tree
+ * @param (n) Nonterminal ID
+ * @param (ctx) Context
+ * @return std::nullopt if it cannot find parents, otherwise RecursionInfo
+ */
+std::optional<RecursionInfo> RecursionInfo::New(Tree& t,
+                                                const NTermID& n,
+                                                Context& ctx) {
+  std::unordered_map<NodeID, NodeID> recursive_parents;
+  std::vector<NodeID> node_by_offset;
+  std::vector<size_t> depth_by_offset;
+
+  std::optional<Parent> r = RecursionInfo::FindParents(t, n, ctx);
+  if (!r) {
+    return std::nullopt;
+  }
+
+  std::tie(recursive_parents, node_by_offset, depth_by_offset) = r.value();
+
+  auto sampler = WalkerDiscreteDistribution<size_t>(depth_by_offset);
+
+  return RecursionInfo(std::move(recursive_parents),
+                       std::move(sampler),
+                       std::move(depth_by_offset),
+                       std::move(node_by_offset));
+}
+
+/**
+ * @fn
+ * @brief Find parents
+ * @params (t) Tree
+ * @params (nt) Nonterminal ID
+ * @params (ctx) Context
+ * @return A tuple of NodeID->NodeID map, vector of NodeID, and vector of depth (std::optional)
+ */
+std::optional<Parent> RecursionInfo::FindParents(
+  Tree& t, const NTermID& nt, Context& ctx
+) {
+  using Stack = std::pair<std::optional<NodeID>, size_t>;
+  std::vector<Stack> stack;
+  stack.emplace_back(std::nullopt, 0);
+
+  std::optional<Parent> res = std::nullopt;
+
+  for (size_t i = 0; i < t.rules().size(); i++) {
+    NodeID node(i);
+    auto [maybe_parent, depth] = stack.back();
+    stack.pop_back();
+
+    if (ctx.GetNT(t.rules()[i]) == nt) {
+      if (maybe_parent) {
+        std::unordered_map<NodeID, NodeID> parents;
+        std::vector<NodeID> ids;
+        std::vector<size_t> weights;
+        if (res)
+          std::tie(parents, ids, weights) = res.value();
+
+        parents[node] = maybe_parent.value();
+        ids.emplace_back(node);
+        weights.emplace_back(depth);
+
+        res = Parent(parents, ids, weights);
+      }
+
+      maybe_parent = node;
+    }
+
+    for (size_t j = 0; j < ctx.GetNumChildren(t.rules()[i]); j++) {
```

Comment:
`ctx.GetNumChildren(t.rules()[i])` will be evaluated in each iteration. It should be moved outside of the loop.

Reply:
Fixed in [82c195c](https://github.com/fuzzuf/fuzzuf/pull/23/commits/82c195c436c8cfb04c86daa5abc249ced0e6b126)